app/utils/client-hints.tsx

```diff
@@ -0,0 +1,139 @@
+    .find(c => c.startsWith(`${hint.cookieName}=`))
+    ?.split('=')[1]
+
+  return value ? decodeURIComponent(value) : null
+}
+
+/**
+ *
+ * @param request {Request} - optional request object (only used on server)
+ * @returns an object with the client hints and their values
+ */
+export function getHints(request?: Request) {
+  let cookieString = ''
+  if (typeof document !== 'undefined') {
+    cookieString = document.cookie
+  } else if (typeof request !== 'undefined' && request.headers.has('Cookie')) {
+    cookieString = request.headers.get('Cookie')!
+  }
+
+  return Object.entries(clientHints).reduce(
+    (acc, [name, hint]) => {
+      const hintName = name as ClientHintNames
+      if ('transform' in hint) {
+        acc[hintName] = hint.transform(
+          getCookieValue(cookieString, hintName) ?? hint.fallback,
+        )
+      } else {
+        // @ts-expect-error - this is fine (PRs welcome though)
+        acc[hintName] = getCookieValue(cookieString, hintName) ?? hint.fallback
+      }
+      return acc
+    },
+    {} as {
+      [name in ClientHintNames]: (typeof clientHints)[name] extends {
+        transform: (value: any) => infer ReturnValue
+      }
+        ? ReturnValue
+        : (typeof clientHints)[name]['fallback']
+    },
+  )
+}
+
+/**
+ * @returns an object with the client hints and their values
+ */
+export function useHints() {
+  const requestInfo = useRequestInfo()
+  return requestInfo.hints
+}
+
+/**
+ * @returns inline script element that checks for client hints and sets cookies
+ * if they are not set then reloads the page if any cookie was set to an
+ * inaccurate value.
+ */
+export function ClientHintCheck({nonce}: {nonce: string}) {
+  const {revalidate} = useRevalidator()
+  React.useEffect(() => {
+    const themeQuery = window.matchMedia('(prefers-color-scheme: dark)')
+    function handleThemeChange() {
+      document.cookie = `${clientHints.theme.cookieName}=${
+        themeQuery.matches ? 'dark' : 'light'
+      }`
+      revalidate()
+    }
+    themeQuery.addEventListener('change', handleThemeChange)
+    return () => {
+      themeQuery.removeEventListener('change', handleThemeChange)
+    }
+  }, [revalidate])
+
+  return (
+    <script
+      nonce={nonce}
+      dangerouslySetInnerHTML={{
+        __html: `
+const cookies = document.cookie.split(';').map(c => c.trim()).reduce((acc, cur) => {
+	const [key, value] = cur.split('=');
+	acc[key] = value;
+	return acc;
+}, {});
+let cookieChanged = false;
+const hints = [
+${Object.values(clientHints)
+  .map(hint => {
+    const cookieName = JSON.stringify(hint.cookieName)
+    return `{ name: ${cookieName}, actual: String(${hint.getValueCode}), cookie: cookies[${cookieName}] }`
+  })
+  .join(',\n')}
+];
+for (const hint of hints) {
+	if (decodeURIComponent(hint.cookie) !== hint.actual) {
+		cookieChanged = true;
+		document.cookie = encodeURIComponent(hint.name) + '=' + encodeURIComponent(hint.actual) + ';path=/';
```

Comment:
I've made a few to this that I'd like applied here: https://github.com/epicweb-dev/epic-stack/commit/3dca3b0dffc1dc84a77e0cdbb306f9397e5a0340